src/main/java/com/redhat/jenkins/plugins/cachet/pipeline/CachetGatingWaitForResourcesStep.java

```diff
@@ -0,0 +1,163 @@
+
+    private final List<String> resources;
+
+    @DataBoundConstructor
+    public CachetGatingWaitForResourcesStep(List<String> resources) {
+        this.resources = new ArrayList<>(resources);
+    }
+
+    public List<String> getResources(){
+        return resources;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(resources, context);
+    }
+
+    public static class Execution extends SynchronousStepExecution<Map<String, Long>> {
+
+        private transient final List<String> resources;
+
+        private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
+        static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
+        static final long MAX_RECURRENCE_PERIOD = 15000; // ¼min
+        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+        Execution(List<String> resources, StepContext context) {
+            super(context);
+            this.resources = resources;
+        }
+
+        @Override
+        protected Map<String, Long> run() throws Exception {
+            Map<String, Long> waitForResourcesMet = new HashMap<>();
+            List<String> resourcesToWaitFor = new ArrayList<>();
+
+            Map<String, CachetGatingMetrics> metrics = getCurrentMetrics();
+
+            for (String resource : resources){
+                if(!metrics.containsKey(resource))
+                    throw new AbortException("One or more resources you want to wait for does not exist.");
+
+                CachetGatingMetrics tmpResourceMet = metrics.get(resource);
+                if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                    waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                }
+                else {
+                    resourcesToWaitFor.add(resource);
+                }
+            }
+            
+            while (!resourcesToWaitFor.isEmpty()){// So there are resources we need to wait for
+                sleepingHandler(resourcesToWaitFor);
+                for (String resource : resourcesToWaitFor){
+                    CachetGatingMetrics tmpResourceMet = getCurrentMetrics().get(resource);
+                    if(tmpResourceMet.getGatingStatus().equals("Operational")){
+                        resourcesToWaitFor.remove(resource);
+                        waitForResourcesMet.put(resource, tmpResourceMet.getGatedTimeElapsed());
+                    }
+                }
+            }
+          return waitForResourcesMet;
+        }
+
+        /**
+         * Help method
+         * @return The current metrics
+         * @throws Exception
+         */
+        public Map<String, CachetGatingMetrics> getCurrentMetrics() throws Exception {
+            Run run = getContext().get(Run.class);
+            Action action = run.getAction(com.redhat.jenkins.plugins.cachet.CachetGatingAction.class);
+            if (action == null) {
+                action = new CachetGatingAction();
```

Comment:
I am not sure there is a situation where there is no action attached, but pretending it is this way will cause the metrics to be empty and it will blow up during the step execution with `AbortException`, no? Following the good style I wonder if there is a reason for the action to have an empty constructor. What is the value of the action if it has no data?